subsys/usb/class/loopback.c

```diff
@@ -0,0 +1,156 @@
+/*
+ * USB loopback function
+ *
+ * Copyright (c) 2018 Phytec Messtechnik GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <init.h>
+
+#include <usb/usb_device.h>
+#include <usb/usb_common.h>
+#include <usb_descriptor.h>
+
+#define SYS_LOG_LEVEL SYS_LOG_LEVEL_ERROR
+#define SYS_LOG_DOMAIN "usb/loopback"
+#include <logging/sys_log.h>
+
+#if !defined(CONFIG_USB_COMPOSITE_DEVICE)
+static u8_t interface_data[64];
+#endif
+
+static u8_t loopback_buf[1024];
+
+static void loopback_bulk_out_cb(u8_t ep,
+				 enum usb_dc_ep_cb_status_code ep_status)
+{
+	u32_t bytes_to_read;
+
+	usb_read(ep, NULL, 0, &bytes_to_read);
+	SYS_LOG_DBG("ep 0x%x, bytes to read %d ", ep, bytes_to_read);
+	usb_read(ep, loopback_buf, bytes_to_read, NULL);
```

Comment:
should it send data back to IN?

Reply:
IN endpoint (buffer) is always prepared for the IN token and independent of OUT endpoint.